src/lib/operators/window_function_evaluator/window_function_evaluator_traits.hpp

```diff
@@ -0,0 +1,319 @@
+      return NonNullableImpl::node_from_value(*input);
+    return std::nullopt;
+  }
+
+  static TreeNode combine(TreeNode lhs, TreeNode rhs) {
+    if (!lhs && !rhs)
+      return std::nullopt;
+    return NonNullableImpl::combine(lhs.value_or(NonNullableImpl::neutral_element),
+                                    rhs.value_or(NonNullableImpl::neutral_element));
+  }
+
+  static OutputType transform_query(TreeNode node) {
+    if (node)
+      return NonNullableImpl::transform_query(*node);
+    return std::nullopt;
+  }
+};
+
+//////////////////////////////////////////////////////////////////////////////
+
+struct AllRanksOnePassBase {
+  struct State {
+    const RelevantRowInformation* previous_row;
+    uint64_t rank;
+    uint64_t dense_rank;
+    uint64_t row_number;
+  };
+
+  static State initial_state(const RelevantRowInformation& row) {
+    return State{
+        .previous_row = &row,
+        .rank = 1,
+        .dense_rank = 1,
+        .row_number = 1,
+    };
+  }
+
+  static void update_state(State& state, const RelevantRowInformation& row) {
+    ++state.row_number;
+    if (!state.previous_row->is_peer_of(row)) {
+      state.rank = state.row_number;
+      ++state.dense_rank;
+    }
+    state.previous_row = &row;
+  }
+};
+
+};  // namespace impls
+
+//////////////////////////////////////////////////////////////////////////////
+
+template <typename InputColumnTypeT, WindowFunction window_function>
+struct WindowFunctionEvaluatorTraits {};
+
+//////////////////////////////////////////////////////////////////////////////
+
+template <typename InputColumnTypeT>
+struct WindowFunctionEvaluatorTraits<InputColumnTypeT, WindowFunction::Rank> {
+  using InputColumnType = InputColumnTypeT;
+  using OutputColumnType = typename WindowFunctionTraits<InputColumnType, WindowFunction::Rank>::ReturnType;
+
+  struct OnePassImpl : impls::AllRanksOnePassBase {
+    static OutputColumnType current_value(const State& state) {
+      return state.rank;
+    }
+  };
+};
+
+template <typename InputColumnTypeT>
+struct WindowFunctionEvaluatorTraits<InputColumnTypeT, WindowFunction::DenseRank> {
+  using InputColumnType = InputColumnTypeT;
+  using OutputColumnType = typename WindowFunctionTraits<InputColumnType, WindowFunction::DenseRank>::ReturnType;
+
+  struct OnePassImpl : impls::AllRanksOnePassBase {
+    static OutputColumnType current_value(const State& state) {
+      return state.dense_rank;
+    }
+  };
+};
+
+template <typename InputColumnTypeT>
+struct WindowFunctionEvaluatorTraits<InputColumnTypeT, WindowFunction::RowNumber> {
+  using InputColumnType = InputColumnTypeT;
+  using OutputColumnType = typename WindowFunctionTraits<InputColumnType, WindowFunction::RowNumber>::ReturnType;
+
+  struct OnePassImpl : impls::AllRanksOnePassBase {
+    static OutputColumnType current_value(const State& state) {
+      return state.row_number;
+    }
+  };
+};
+
+//////////////////////////////////////////////////////////////////////////////
```

Comment:
General question: Why are you using these slashes? You could consider using regions instead.

Reply:
Regions are not used, so maybe this don't make sense. We don't have a better idea so is also ok to use it.